### PR TITLE
Users/andreas/fix midi without instruments

### LINKED ISSUE #9: App crashes when midi file doesn't specify an instrument
When loading a mii file not containing a ProgramChangedEvent the app crashes. Instead Coimbra should use the default midi instrument.

### PR DESCRIPTION
**Pull request by andreasbalzer**
Fixes #9 so that midi files without a ProgramChangedEvent specifying an instrument can be loaded and played.

### CODE REVIEW

File: ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs [hunk 1]
```diff
@@ -241,18 +241,37 @@ private static string GetTrackDisplayName(StorageFile file, MidiFile midi)
...
+
+                if (currentEvent is ProgramChangeEvent currentParsedEvent)
+                {
+                    programNumber = currentParsedEvent.ProgramNumber;
+                }
+
+                if (this.Instruments.ContainsKey(currentEvent.Channel))
                 {
-                    if (!this.Instruments[programChangeEvent.Channel].Contains(programChangeEvent.ProgramNumber))
+                    if (!this.Instruments[currentEvent.Channel].Contains(programNumber))
```
Comment (muiriswoulfe) [inline ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs:266]: If we don't want duplicates, a `HashSet` would likely have been better than a `List`. I realise you didn't change this now, but calling this out as I see it. Note, by the way, that the ordering in a `HashSet` can change so if you want the preserve the ordering, the `List` is the right data structure.

File: ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs [hunk 2]
```diff
@@ -241,18 +241,37 @@ private static string GetTrackDisplayName(StorageFile file, MidiFile midi)
 
         private void AddInstruments(MidiFile midi)
         {
-            foreach (var programChangeEvent in midi.GetTimedEvents().Select(e => e.Event).OfType<ProgramChangeEvent>())
+            var timedEvents = midi.GetTimedEvents();
+            var programChangedEvents = timedEvents.Select(e => e.Event).OfType<ProgramChangeEvent>().ToList<ChannelEvent>();
```
Comment (muiriswoulfe) [inline ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs]: It looks like you don't use `programChangedEvents` in many places. I believe you could skip this entirely and just apply the result of this operation to `events`.

**Review by muiriswoulfe — APPROVED** [review]

File: ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs (hunk 1 above)
Comment (andreasbalzer) [reply ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs:266]: I think maintaining the order of instruments is useful. We might want to extend this in the future to split different instruments per channel into different players too.

**Review by muiriswoulfe — APPROVED** [review]

File: ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs (hunk 1 above)
Comment (muiriswoulfe) [reply ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs:266]: Makes sense. Thanks for clarifying.

**Pull request merged**